File: src/my_py_pkg/my_py_pkg/add_two_ints_client_no_oop.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node

from example_interfaces.srv import AddTwoInts
 
def main(args=None):
    rclpy.init(args=args)
    
    node = Node("add_two_ints_no_oop")
    client = node.create_client(AddTwoInts, "add_two_ints")

    # Wait for server to be up and running
    # Treat the returned boolean of client.wait_for_service(wait_frequency) as a boolean returned by is_server_up(wait_frequency)
    while not client.wait_for_service(1.0):
        node.get_logger().warn("Waiting for service add_two_ints...")
    
    # Create service request
    req = AddTwoInts.Request()
    req.a = 26
    req.b = 43

    # Request service
    # Returned object is of asyncio.Future type
    # The response from server can be extracted using asyncio.Future().result()
    future = client.call_async(req)

    # Keep the node alive until the asyncronous action is complete
    rclpy.spin_until_future_complete(node, future)

    # Check if the service succeeded or failed
    try:
        res = future.result()
        node.get_logger().info(f'{req.a} + {req.b} = {res.sum}')
    except Exception as e:
        node.get_logger().error(f'Service failed: {e:%r}')

    rclpy.shutdown()
 
 
if __name__ == "__main__":
    main()

File: src/my_cpp_pkg/src/add_two_ints_client_no_oop.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"
 
int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);

    auto node = std::make_shared<rclcpp::Node>("add_two_ints_client_no_oop");
    auto client = node->create_client<example_interfaces::srv::AddTwoInts>("add_two_ints");

    // Wait for server to be up and running
    while(!client->wait_for_service(std::chrono::seconds(1)))
    {
        RCLCPP_WARN(node->get_logger(), "Waiting for add_two_ints server to start...");
    }

    // Create a request
    auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
    // auto request = example_interfaces::srv::AddTwoInts::Request::SharedPtr(); -> DOES NOT WORK! Gives segmentation fault!!
    request->a = 20;
    request->b = 10;

    auto future = client->async_send_request(request);
    if (future.valid()) {
        if (rclcpp::spin_until_future_complete(node, future) == rclcpp::FutureReturnCode::SUCCESS) {
            auto sum = future.get()->sum;
            RCLCPP_INFO(node->get_logger(), "%ld + %ld = %ld", request->a, request->b, sum);
        } else {
            RCLCPP_ERROR(node->get_logger(), "Service request for add_two_ints server FAILED!");
        }
    } else {
        RCLCPP_ERROR(node->get_logger(), "Failed to send request to add_two_ints server!");
    }

    rclcpp::shutdown();
    return 0;
}